### DISCUSSION TITLE
How to wait all data to be sent on stream before main thread end program

### DISCUSSION THREAD

**Discussion by ami-GS** [microsoft/msquic#2207] [Q&A] [answered]
Some application just send data once then quickly close by design. and launched by cron etc.
How can the app over quic wait all the data on stream to be sent?
Even if the app call *Shutdown(), these are non-blocking functions, app could finish before the data sent.

**Comment by nibanks** [comment, marked as answer]
There are a couple ways you might achieve this, but all of them rely on either explicitly or implicitly waiting on the "shutdown complete" event before closing the object(s). The simplest way for a fire-and-forget style protocol would be to so something like the [sample](https://github.com/microsoft/msquic/blob/main/src/tools/sample/sample.c). You will notice the `RunClient` simply starts the connection async and then returns, eventually calling `RegistrationClose`. The start will asynchronously kick off other work, like opening and sending stream data. The main thread that calls `RegistrationClose` will block (wait) until the connection has been closed; so as long as the code cleans up asynchronously, correctly, eventually everything completes nicely and then returns/exits.

**Comment by ami-GS** [comment]
Thanks! I did not notice the blocking call!
